utils/pattern/src/frontend/mod.rs

```diff
@@ -9,17 +9,57 @@ mod serde;
 use crate::common::*;
 use crate::Error;
 use crate::PatternOrUtf8Error;
+use crate::SinglePlaceholderPattern;
 #[cfg(feature = "alloc")]
 use crate::{Parser, ParserOptions};
 #[cfg(feature = "alloc")]
 use alloc::{borrow::ToOwned, str::FromStr, string::String};
+use core::ops::Deref;
 use core::{
     convert::Infallible,
     fmt::{self, Write},
     marker::PhantomData,
 };
 use writeable::{adapters::TryWriteableInfallibleAsWriteable, PartsWrite, TryWriteable, Writeable};
 
+#[allow(clippy::unwrap_used)]
+#[cfg(feature = "alloc")]
+impl SinglePlaceholderPattern<str> {
+    pub fn encode_store(prefix: &str, suffix: &str) -> alloc::boxed::Box<str> {
+        alloc::format!(
+            "{}{prefix}{suffix}",
+            char::from_u32(prefix.len() as u32 + 1).unwrap()
+        )
+        .into_boxed_str()
+    }
+}
```

Comment:
Issue: I would prefer to avoid adding methods directly to `SinglePlaceholderPattern<str>`. Methods should be added to `Pattern<Store>`.

Besides that, this code duplicates logic that lives in the pattern constructor. You should build the pattern using the standard APIs and then use `take_store`.

The constructor you probably want is `Pattern::try_from_items`.